syntax="proto3";

package descarteslabs.workflows;

import "google/protobuf/descriptor.proto";


extend google.protobuf.FieldOptions {
  string mimetype = 3226;
}


// NOTE: We use a 'not_' naming convention for boolean fields
// whose default values are True. The default value of a boolean field in
// proto3 is False; this convention allows us properly handle boolean fields with
// a default value of True. Applications presenting these fields to
// users should do so without the 'not_' prefix and with a default of True.
// Mimetypes support passing these parameters without the 'not_' prefix,
// in which case the value is inverted. 

message Format {
  // TODO: add oneof for these once we can handle it better on the backend
  PyArrow pyarrow = 1 [(mimetype) = "application/vnd.pyarrow"];
  JSON json = 2 [(mimetype) = "application/json"];
  GeoJSON geojson = 3 [(mimetype) = "application/geo+json"];
  CSV csv = 4 [(mimetype) = "text/csv"];
  PNG png = 5 [(mimetype) = "image/png"];
  GeoTIFF geotiff = 6 [(mimetype) = "image/tiff"];

  bool has_pyarrow = 20;
  bool has_json = 21;
  bool has_geojson = 22;
  bool has_csv = 23;
  bool has_png = 24;
  bool has_geotiff = 25;
}

message PyArrow {
  enum Compression {
    lz4 = 0;
    brotli = 1;
  }
  Compression compression = 1;
}

message JSON {}

message GeoJSON {}

message CSV {}

message PNG {}

message GeoTIFF {
  bool not_overviews = 1;
  bool not_tiled = 2;
  enum Compression {
    LZW = 0;
    None = 1;
    JPEG = 2;
    JPEG2000 = 3;
  }
  Compression compression = 3;
}
